src/snowflake/connector/query_context_cache.py

```diff
@@ -0,0 +1,271 @@
+from hashlib import md5
+from io import BytesIO
+from logging import DEBUG, getLogger
+from threading import Lock
+
+from sortedcontainers import SortedSet
+
+from .options import pyarrow as pa
+
+logger = getLogger(__name__)
+
+
+@total_ordering
+class QueryContextElement:
+    def __init__(self, id: int, read_timestamp: int, priority: int, context: bytearray):
+        self._id = id
+        self._read_timestamp = read_timestamp
+        # priority values are 0..N with 0 being the highest priority
+        self._priority = priority
+        self._context = context
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    @property
+    def read_timestamp(self) -> int:
+        return self._read_timestamp
+
+    @read_timestamp.setter
+    def read_timestamp(self, timestamp: int) -> None:
+        self._read_timestamp = timestamp
+
+    @property
+    def priority(self) -> int:
+        return self._priority
+
+    @property
+    def context(self) -> bytearray:
+        return self._context
+
+    @context.setter
+    def context(self, ctx: bytearray) -> None:
+        self._context = ctx
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            return False
+        return (
+            self._id == other.id
+            and self._read_timestamp == other.read_timestamp
+            and self._priority == other.priority
+            and self._context == other.context
+        )
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            raise NotImplementedError(
+                f"cannot compare QueryContextElement with object of type {type(other)}"
+            )
+        return self._priority < other.priority
+
+    def __hash__(self) -> int:
+        _hash = 31
+
+        _hash = _hash * 31 + self._id
+        _hash += (_hash * 31) + self._read_timestamp
+        _hash += (_hash * 31) + self._priority
+        if self._context:
+            _hash += (_hash * 31) + int.from_bytes(md5(self._context).digest(), "big")
+        else:
+            _hash += _hash * 31
+        return _hash
+
+    def __str__(self) -> str:
+        return f"({self._id}, {self._read_timestamp}, {self._priority})"
+
+
+class QueryContextCache:
+    QUERY_CONTEXT_SCHEMA = pa.schema(
+        [
+            pa.field("id", pa.int64(), nullable=False),
+            pa.field("timestamp", pa.int64(), nullable=False),
+            pa.field("priority", pa.int64(), nullable=False),
+            pa.field("context", pa.binary(), nullable=True),
+        ]
+    )
+
+    def __init__(self, capacity: int):
+        self._capacity = capacity
+        self._id_map: dict[int, QueryContextElement] = {}
+        self._priority_map: dict[int, QueryContextElement] = {}
+        # stores elements sorted by priority. Element with
+        # least priority value has the highest priority
+        self._treeset: set[QueryContextElement] = SortedSet()
```

Comment:
I'm happy to update the code using by using `key` for `SortedSet` but I already have an implementation of `__lt__` and `__eq__` for `QueryContextElement` based on @sfc-gh-igarish's doc: https://docs.google.com/document/d/1OOJE0lYFDkjkGoQFrnH28dyp7Eve9k8faa69Qt1OLtc/edit#heading=h.uj8lyr9wa82r. I believe `__lt__` offers more flexible comparison in this case because it can be modified to have more complex comparisons vs `key`. Do you agree?

Reply:
Ah I see. I'm not aware of jdbc's existing implementation. I still think a custom comparator is better (e.g. using `cmp_to_key` to convert a comparator to key), but given JDBC's choice, I don't have a strong opinion.